smtlib-backends-process/tests/Examples.hs

```diff
@@ -52,35 +54,35 @@ setOptions =
             Process.reportError = LBS.putStr . (`LBS.snoc` '\n')
           }
    in Process.with myConfig $ \handle -> do
-        -- since the 'Process' module exposes its 'Handle' datatype entirely, we can also
-        -- change the settings of the underlying process
-        -- you probably won't need to do this as the library already choose these
-        -- settings to ensure the communication with the solvers is as fast as
-        -- possible
-        let p = Process.process handle
-            stdin = getStdin p
-        -- for instance here we change the buffering mode of the process' input channel
-        hSetBuffering stdin LineBuffering
-        -- we can then use the backend as before
-        let backend = Process.toBackend handle
-        solver <- initSolver Queuing backend
+        solver <- initSolver Queuing $ Process.toBackend handle
         _ <- command solver "(get-info :name)"
         return ()
 
--- | An example of how to close the 'Process' backend's underlying process manually,
--- instead of relying on 'Process.with' or 'Process.close'.
-manualExit :: IO ()
-manualExit = do
-  -- launch a new process with 'Process.new'
-  handle <- Process.new Process.defaultConfig
-  -- do some stuff
-  doStuffWithHandle handle
-  -- kill the process with 'Process.kill'
-  -- other options include using 'Process.close' to ensure the process exits
-  -- gracefully
-  Process.kill handle
-  where
-    doStuffWithHandle _ = return ()
+-- | An example of how to get the backend's underlying process and manage it
+-- manually.
+underlyingProcess :: IO ()
+underlyingProcess = do
+  -- since the 'Process' module exposes its 'Handle' datatype entirely, we have
+  -- direct access to the process and its I/O channels
+  -- you'll probably never need this as the library already provides enough
+  -- bindings for common needs and chooses the process' settings to ensure the
+  -- communication with solvers is as fast as possible
+  --
+  -- we'll close the process manually so we just launch it with 'Process.new'
+  -- instead of using `Process.with`
+  Process.Handle {..} <- Process.new Process.defaultConfig
+  -- we can now change the settings of the underlying process
+  -- for instance here we change the buffering mode of the process' input channel
+  hSetBuffering hIn LineBuffering
+  -- the default way to close the process is to use 'Process.close', which sends
+  -- it a SIGTERM
+  -- if you don't like this, you can always send it an @(exit)@ command and wait
+  -- for it to end, but make sure you also release its other resources
+  LBS.hPutStrLn hIn "(exit)"
+  cancel errorReader
```

Comment:
```suggestion
  -- for it to end, but make sure you also release its IO handles
  LBS.hPutStrLn hIn "(exit)"
```
It wouldn't be necessary to cancel the `errorReader`. It should stop when it either finds the `hErr` handle has been closed or when it receives EOF as a result of terminating the process.

Reply:
Should we also remove this call from `Process.close` then ?

Reply:
:thinking: there is one possible strange behavior where the `errorReader` could still deliver one last message after the process is terminated, and before it tries to read from the `hErr` handle.

Do you think it is necessary to make the user aware of the errorReader thread and of this possibility?

Reply:
Not really no, and we can always add the option back if someone complains.

So I should remove the `errorReader` field entirely then, right ?

Reply:
Will GHC optimize the async thread for the `errorReader` when the logging function is `void` though ? If not we probably want to leave the option to the user to cancel this thread, or better yet not run the thread at all.

Reply:
I'm trying to think if we have some alternative to using a thread.

One option that I'm considering is to make the client responsible from reading from the `hErr` handle if he cares.

Reply:
> One option that I'm considering is to make the client responsible for reading from the hErr handle if she cares.

We could go with this approach and also add as a configuration option the `StdStream` spec to use for stderr.

Reply:
I implemented logging at first to be as close to SimpleSMT as possible. Now that we're using process we're not logging possible failures when closing the process' handles, so the only use of the `reportError` function is for this `errorReader`.

I think it makes sense to keep `reportError` if we use it in multiple places. If we only use it for `errorReader` specifying `reportError` will likely cost as much to the user as them manually implementing `errorReader`, or whatever option they choose to go with.

Reply:
https://github.com/tweag/smtlib-backends/pull/55/commits/d309d6cc09298cd575020c791b7affa7e0098f1f

Reply:
Yes, this is for another PR.

> also add the corresponding options for stdin and stdout

Or just take a `CreateProcess` value as argument.